addons/purchase_requisition/models/purchase.py

```diff
@@ -127,6 +128,9 @@ def button_confirm(self):
 
     @api.model_create_multi
     def create(self, vals_list):
+        if self.env.context.get('vendor_required', False):
+            if any(not vals.get('partner_id', False) for vals in vals_list):
+                raise UserError(_('A vendor is required on the alternative purchase order(s).'))
```

Comment:
I had to add the context value for creating an RFQ from the ["Link to Existing RfQ"](https://github.com/odoo/odoo/blob/master/addons/purchase_requisition/views/purchase_views.xml#L34-L36)) button, not for ["Create Alternative"](https://github.com/odoo/odoo/blob/master/addons/purchase_requisition/views/purchase_views.xml#L27) button above it. `origin_po_id` is passed indeed when using the latter, but for the former it is not. I also tried to debug the `create` to see if I can catch it there (if `purchase_group_id` or `alternative_po_ids` are passed) but they were not. The link is done in the `write` of the original PO. We can catch it there and prevent the linking, but the problem is that the other PO would have been created already (for no use). What do you think then, should we keep the context or catch in `write` or even a better idea 😅?

Regarding the other scenario, I already [added a domain tuple](https://github.com/odoo/odoo/pull/167179/files#diff-e0552765f0f0e122b0a8ecee37a4070d5b55eb944131b3cb67e05398457f208cR33) to `alternative_po_ids` for this case, so that those RFQs without a vendor won't show up in the wizard and users cannot select them. Do you think we should also double-check in `write`?

Reply:
But when you `Link to Existing RfQ`, you're not creating a new PO, you're linking to an existing one. If I remember correctly, the xml says `create`, but that's just the element type for the "Add" (i.e. link to an existing record) for Many2Many fields. That override is just to change what is displayed there.

So indeed, you can add a check into the write (if you want the assignment to completely avoided even in the rpc/customized cases), but it still seems unnecessarily restrictive/confusing to me to make it so you can't link 2 POs that both don't have vendors assigned to me. Your domain also won't stop someone from creating a no vendor PO (that isn't created from a MTO/Dropship, i.e. has no linked SO) and linking that to other POs with vendors.

Reply:
Actually `Link to Existing RfQ` is used for both, linking and creating new ones:
<img width="983" alt="Screenshot 2024-06-26 at 3 12 34 PM" src="https://github.com/odoo/odoo/assets/60348524/b12234ad-0aad-4dc6-8806-279be5953eb0">
This is the wizard that opens when creating on it. And when clicking on `New` button below, a new wizard opens with the form of purchase order.

But I added a [check to the `write`](https://github.com/odoo/odoo/pull/167179/files#diff-e0552765f0f0e122b0a8ecee37a4070d5b55eb944131b3cb67e05398457f208cR164-R166) too for RPCs or any actions not done through the web client.

Reply:
Yes, it does